### ISSUE TITLE
Cannot use DHCP and IPv6 SLAAC in vnet jail

### ISSUE THREAD

**Issue by michael-o** [cbsd/cbsd#540]
Mandatory info for bug reports:

FreeBSD version ( uname -a ): 12.2-STABLE r365640 GENERIC

CBSD version ( cbsd version ): 12.2.0

I have created a vnet jail and set ip4_addr to `0`. In the jail I have set:
```
ifconfig_eth0="DHCP"
ifconfig_eth0_ipv6="inet6 -no_radr accept_rtadv"
rtsold_enable="YES"
```

The startup of the jail goes from seconds to minutes.:
```
ip not identified: 0
ip not identified: 0
ip not identified: 0
No arg for -p option
Stoping jail: pg1, parallel timeout=5
Destroy epair: epair1a
jstop done in 4 seconds
ip not identified: 0
create epair: epair1:em0 epair1b: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=8<VLAN_MTU>
        ether 02:ff:f0:1e:0c:0e
        hwaddr 02:5e:be:c4:88:0b

ip not identified: 0
No arg for -p option
ip not identified: 0
set resource limit: [ ]
jail renice: 1
Applying /usr/ports/CHANGES -A28 ^20140930 work-around
Starting jail: pg1, parallel timeout=5
pg1: created
eth0
late_start in progress...
ELF ldconfig path: /lib /usr/lib /usr/lib/compat /usr/local/lib /usr/local/lib/compat/pkg /usr/local/lib/compat/pkg /usr/local/lib/perl5/5.32/mach/CORE
32-bit compatibility ldconfig path: /usr/lib32
Setting hostname: pg1.fritz.box.
Starting Network: lo0 eth0.
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
        inet 127.0.0.1 netmask 0xff000000
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
eth0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=8<VLAN_MTU>
        ether 02:ff:f0:1e:0c:0e
        hwaddr 02:5e:be:c4:88:0b
        inet6 fe80::ff:f0ff:fe1e:c0e%eth0 prefixlen 64 scopeid 0x2
        groups: epair
        media: Ethernet 10Gbase-T (10Gbase-T <full-duplex>)
        status: active
        nd6 options=23<PERFORMNUD,ACCEPT_RTADV,AUTO_LINKLOCAL>
add host 127.0.0.1: gateway lo0 fib 0: route already in table
add host ::1: gateway lo0 fib 0: route already in table
add net fe80::: gateway ::1
add net ff02::: gateway ::1
add net ::ffff:0.0.0.0: gateway ::1
add net ::0.0.0.0: gateway ::1
Waiting 30s for the default route interface: .............................
Starting rtsold.
Creating and/or trimming log files.
Starting syslogd.
Clearing /tmp (X related).
Updating motd:.
Updating /var/run/os-release done.
Starting sendmail_submit.
Starting sendmail_msp_queue.

Starting cron.

Sa.  7 Nov. 2020 20:09:50 CET
jstart done in 1 minutes and 54 seconds
jrestart done in 1 minutes and 58 seconds
jrestart done in 1 minutes and 58 seconds
.
```

and in the jail:
```
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> metric 0 mtu 16384
        options=680003<RXCSUM,TXCSUM,LINKSTATE,RXCSUM_IPV6,TXCSUM_IPV6>
        inet6 ::1 prefixlen 128
        inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
        inet 127.0.0.1 netmask 0xff000000
        groups: lo
        nd6 options=21<PERFORMNUD,AUTO_LINKLOCAL>
eth0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> metric 0 mtu 1500
        options=8<VLAN_MTU>
        ether 02:ff:f0:1e:0c:0e
        hwaddr 02:5e:be:c4:88:0b
        inet6 fe80::ff:f0ff:fe1e:c0e%eth0 prefixlen 64 scopeid 0x2
        inet6 PREFIX:ff:f0ff:fe1e:c0e prefixlen 64 autoconf
        groups: epair
        media: Ethernet 10Gbase-T (10Gbase-T <full-duplex>)
        status: active
        nd6 options=23<PERFORMNUD,ACCEPT_RTADV,AUTO_LINKLOCAL>
```

while router solicitation works, IPv4 is unusable:
```
Routing tables

Internet:
Destination        Gateway            Flags     Netif Expire
127.0.0.1          link#1             UH          lo0
```

Setting back to `REALDHCP` while retaining SLAAC startup still takes forever, SLAAC works, no DHCP.

My vnet knowledge is very little, but it seems like even with a virtual net stack I don't have the control I should have, do I?

What am I doing wrong here?

**Comment by mekanix** [comment]
`REALDHCP` should be used only if you have DHCP server for your jails. Unfortunately, I have zero experience with IPv6 so I can't help more.

**Comment by michael-o** [comment]
Yes, I do have a DHCP server for those jails. As far as I understand to the [FreeBSD Forums](https://forums.freebsd.org/threads/jail-vnet-slaac-ipfw.69690/) this can be passed when the epair is created to `ifconfig`. CBSD does not seem to be accomondated for this. I am not an expect on IPv6, but invested countless hours to understand how the new world works. My ISP provides me a /56 prefix which exposes all nodes, if I like to the internet.

**Comment by mekanix** [comment]
I have to learn how to configure DHCP server for IPv6, but it's a pending task for me. As Reggae already comes with [isc-dhcpd](https://github.com/cbsd/reggae/blob/master/templates/dhcpd.conf) and [dhcpcd](https://github.com/cbsd/reggae/blob/master/templates/dhcpcd.conf), it's ready for IPv6 extension. So what I have to figure out is how to run dual stack DHCP server, first. If you have any tips, I'm all ears.

**Comment by michael-o** [comment]
You don't actually need DHCPv6. Prefix delegation with router solicitation paired with neighborhood discovery is sufficient. Base does not contain a DHCPv6 client.

**Comment by mekanix** [comment]
I know, but I'd like to have names registered in local_unbound, so I'd like DHCP.

**Comment by michael-o** [comment]
> 
> 
> I know, but I'd like to have names registered in local_unbound, so I'd like DHCP.

Me to, this is one of the weaknesses of SLAAC. There is an expired RFC proposal which would made this possible, but never passed :-(

**Comment by olevole** [comment]
AFAIK, 'DHCP' for vnet-based jail does not work, since the network in the vnet container is configured by the administrator and it is not limited to one address. REALDHCP just add ifconfig_eth0="DHCP" entries inside jail.

As for IPV6 PD, /usr/local/sbin/dual-dhclient ( from /usr/ports/net/dual-dhclient ) or dhcp6c should work ( net/dhcp6 ).

Unfortunately, I don’t know what we can improve on the CBSD side, just generate a config file and execute the jail utility

**Comment by michael-o** [comment]
Why then can't I pass `inet6 -no_radr accept_rtadv` to CBSD directly to have SLAAC configured? It seems like it wants IP addresses or specific literals like DHCP only?

**Comment by 4xoc** [comment]
I've been using vnet with v6 and SLAAC for some time now so here're my two cents on it. Configuring everything in rc.conf after the jail has been created is the fastest when just having a new jail once in a while. `ifconfig_eth0_ipv6="inet6 accept_rtadv"` does the trick for me without need for additional tools like rtsold or any dhcpv6 client. To automate this for new jails it should be possible to simply create a custom jail profile that references a custom `jailskeldir` which in turn has a rc.conf with the mentioned content.

**Comment by olevole** [comment]
Hmm, I think we can easily add 'sysrc' for `jcreate` script for any customization, e.g:
```
cbsd jcreate jname=test vnet=1 runasap=1 sysrc="ifconfig_eth0_ipv6+='inet6 accept_rtadv'"
```

or something like that. need to think about how to write it more beautifully for user.

**Comment by michael-o** [comment]
I would prefer cbsd not to fiddle with network config on vnet jails and leave this to the admin. I understand that cbsd fiddles on shared interaces (non-vnet).

**Comment by Peter2121** [comment]
> I would prefer cbsd not to fiddle with network config on vnet jails and leave this to the admin. I understand that cbsd fiddles on shared interaces (non-vnet).

There are use-cases where CBSD should configure VNET jails. For example, in case of multiple VNET jails, the configuration of jails default routers should be done by CBSD at the same time it configures host to get up the GW addresses. If it configures jails IP at the same time - it's even better ;)

**Comment by olevole** [comment]
CBSD does not manage **ifconfig_eth0=** inside a vnet based jail when ip4_addr='0' (disabled). But yes, there are many cases where it might be necessary. For example, when there is a VPC (ipsec, vxlan, wireguard, ...) between several nodes (for example, as described here: https://www.bsdstore.ru/en/13.0.x/wf_vpc_ssi.html). In this case, all jails (bhyve) used the same/shared L2 network (vxlan), but depending on the node, CBSd sets the closest gateway for the container when jail start.

I think there should be both possibilities as a compromise for a wide variety of cases.
Of course we can configure jails in two steps:

1) cbsd jcreate jname=test  ver=13.0 vnet=1 ip4_addr=0 
2) sysrc -qf ~cbsd/jails/jails-data/test-data/etc/rc.conf foo=bar  ( or cbsd sysrc .. )
3) jstart

or via CBSDfile:
```
jail_test()
{
    vnet=1
    ip4_addr="10.0.0.13"
    host_hostname="${jname}.my.domain"
}

postcreate_test()
{
  sysrc ifconfig_eth0+="..."
}
```

but I want to add the ability to customize operations for atomic operation, single command:

```
cbsd jcreate jname=test  ver=13.0 vnet=1 ip4_addr=10.0.0.2/24 runasap=1 sysrc="ifconfig_eth0+='mtu 1450'" ci_gw4=10.0.0.1
```

**Comment by olevole** [comment]
Well, I made a small change (latest available CBSD **13.0.9**) that prevents CBSD from overwriting custom options in jail /etc/rc.conf for ifconfig_eth0. Example:

**case 1**, CBSD does not manage ifconfig_eth0 in /etc/rc.conf ( ip4_addr=0 ):
```
# cbsd jcreate jname=testv6 ver=13.0 vnet=1 ip4_addr=0 runasap=1
# cbsd jexec jname=testv6 grep eth0 /etc/rc.conf
```

**case 2**: CBSD controls ifconfig_eth0 in /etc/rc.conf according to the ip4_addr value, but overwrites only the address, not all parameters:
```
# cbsd jcreate jname=testv6 ver=13.0 vnet=1 ip4_addr=10.0.1.50 ci_gw4=10.0.1.1 runasap=1
# grep -E "^ifconfig|^default" ~cbsd/jails-data/testv6-data/etc/rc.conf
defaultrouter="10.0.1.1"
ifconfig_eth0="10.0.1.50"
```

Now we can add any custom parameters for **ifconfig_eth0** and the CBSD will not overwrite them when the container is restarted:
```
# cbsd sysrc jname=testv6 ifconfig_eth0="inet6 -no_radr accept_rtadv up"
# grep -E "^ifconfig" ~cbsd/jails-data/testv6-data/etc/rc.conf
ifconfig_eth0="inet6 -no_radr accept_rtadv up"

# cbsd jrestart testv6
# grep -E "^ifconfig" ~cbsd/jails-data/testv6-data/etc/rc.conf
ifconfig_eth0="10.0.1.50 inet6 -no_radr accept_rtadv up"
```

Also sysrc was added to jcreate, e.g (see man sysrc):
```
# cbsd jcreate jname=xx ver=13.0 vnet=1 sysrc="ifconfig_eth0+='mtu 1450' inetd_enable=YES" runasap=1
# cbsd jexec jname=xx cat /etc/rc.conf
..
ifconfig_eth0="172.16.0.104 mtu 1450"
inetd_enable="YES"
..
```
